shared/src/commonMain/kotlin/org/hyperskill/app/track_selection/details/presentation/TrackSelectionDetailsFeature.kt

```diff
@@ -0,0 +1,111 @@
+
+import org.hyperskill.app.analytic.domain.model.hyperskill.HyperskillAnalyticEvent
+import org.hyperskill.app.providers.domain.model.Provider
+import org.hyperskill.app.track.domain.model.TrackWithProgress
+
+object TrackSelectionDetailsFeature {
+    internal data class State(
+        val trackWithProgress: TrackWithProgress,
+        val isTrackSelected: Boolean,
+        val isTrackLoadingShowed: Boolean,
+        val contentState: ContentState
+    )
+
+    internal sealed interface ContentState {
+        object Idle : ContentState
+        object Loading : ContentState
+        data class Content(
+            val isFreemiumEnabled: Boolean,
+            val providers: List<Provider>
+        ) : ContentState
+        object NetworkError : ContentState
+    }
+
+    internal fun initialState(
+        trackWithProgress: TrackWithProgress,
+        isTrackSelected: Boolean
+    ) =
+        State(
+            trackWithProgress = trackWithProgress,
+            isTrackSelected = isTrackSelected,
+            isTrackLoadingShowed = false,
+            contentState = ContentState.Idle
+        )
+
+    sealed interface ViewState {
+        object Idle : ViewState
+        object Loading : ViewState
+        data class Content(
+            val title: String,
+            val description: String,
+            val formattedRating: String,
+            val formattedTimeToComplete: String?,
+            val formattedTopicsCount: String,
+            val formattedProjectsCount: String?,
+            val isCertificateAvailable: Boolean,
+            val mainProvider: MainProvider?,
+            val formattedOtherProviders: String?,
+            val isBeta: Boolean,
+            val isCompleted: Boolean,
+            val isSelected: Boolean
+        ) : ViewState
+
+        data class MainProvider(
+            val title: String,
+            val description: String
+        )
+
+        object NetworkError : ViewState
+    }
+
+    sealed interface Message {
+        object Initialize : Message
+
+        object RetryContentLoading : Message
+
+        object SelectTrackButtonClicked : Message
+
+        /**
+         * Analytic
+         */
+        object ViewedEventMessage : Message
+    }
+
+    internal sealed interface FetchProvidersAndFreemiumStatusResult : Message {
+        data class Success(
+            val isFreemiumEnabled: Boolean,
+            val providers: List<Provider>
+        ) : FetchProvidersAndFreemiumStatusResult
+        object Error : FetchProvidersAndFreemiumStatusResult
+    }
+
+    internal sealed interface TrackSelectionResult : Message {
+        object Success : TrackSelectionResult
+        object Error : TrackSelectionResult
+    }
+
+    sealed interface Action {
+        sealed interface ViewAction : Action {
+            sealed interface ShowTrackSelectionStatus : ViewAction {
+                object Success : ShowTrackSelectionStatus
+                object Error : ShowTrackSelectionStatus
+            }
```

Comment:
Let's use the old `ShowTrackSelectionStatus` API (it's more friendly to use in the declarative SwiftUI code):
```kotlin
sealed interface ShowTrackSelectionStatus : ViewAction {
    object Loading : ShowTrackSelectionStatus
    object Error : ShowTrackSelectionStatus
    object Success : ShowTrackSelectionStatus
}
```

Reply:
@ivan-magda 
In my opinion, it's better to hold a state of UI elements that do not disappear after some time (like snackbar on Android) in the state, because UI elements should be hidden at some point in the future by some feature conditions.
Using actions for this purpose we don't know on the platform side whether it should be hidden or not.

Reply:
OK, but on iOS this will lead to some weird behavior - when iOS will handle `ShowTrackSelectionStatus.Error | Success` and show success/error HUD just after that SwiftUI will recompute the layout and `isTrackLoadingShowed == false`, HUD will dismiss without animation.